pkg/table/tables/partition.go

```diff
@@ -1821,31 +1830,66 @@ func (t *partitionTableWithGivenSets) GetAllPartitionIDs() []int64 {
 	return ptIDs
 }
 
+func dataEqRec(loc *time.Location, tblInfo *model.TableInfo, row []types.Datum, rec []byte) (bool, error) {
+	columnFt := make(map[int64]*types.FieldType)
+	for idx := range tblInfo.Columns {
+		col := tblInfo.Columns[idx]
+		columnFt[col.ID] = &col.FieldType
+	}
+	foundData, err := tablecodec.DecodeRowToDatumMap(rec, columnFt, loc)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	for idx, col := range tblInfo.Cols() {
+		if d, ok := foundData[col.ID]; ok {
+			if !d.Equals(row[idx]) {
+				return false, nil
+			}
+		}
+	}
+	return true, nil
+}
+
 // RemoveRecord implements table.Table RemoveRecord interface.
 func (t *partitionedTable) RemoveRecord(ctx table.MutateContext, txn kv.Transaction, h kv.Handle, r []types.Datum, opts ...table.RemoveRecordOption) error {
 	opt := table.NewRemoveRecordOpt(opts...)
 	ectx := ctx.GetExprCtx()
-	pid, err := t.locatePartition(ectx.GetEvalCtx(), r)
+	from, err := t.locatePartition(ectx.GetEvalCtx(), r)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	tbl := t.getPartition(pid)
+	tbl := t.getPartition(from)
 	err = tbl.removeRecord(ctx, txn, h, r, opt)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	if _, ok := t.reorganizePartitions[pid]; ok {
-		pid, err = t.locateReorgPartition(ectx.GetEvalCtx(), r)
-		if err != nil {
+	if _, ok := t.reorganizePartitions[from]; ok {
+		newFrom, err := t.locateReorgPartition(ectx.GetEvalCtx(), r)
+		if err != nil || newFrom == 0 {
 			return errors.Trace(err)
 		}
-		tbl = t.getPartition(pid)
-		err = tbl.removeRecord(ctx, txn, h, r, opt)
+
+		if t.Meta().HasClusteredIndex() {
+			return t.getPartition(newFrom).removeRecord(ctx, txn, h, r, opt)
+		}
+		var found map[string][]byte
+		encodedRecordID := codec.EncodeInt(nil, h.IntValue())
+		newFromKey := tablecodec.EncodeRowKey(newFrom, encodedRecordID)
+
+		found, err = txn.BatchGet(context.Background(), []kv.Key{newFromKey})
```

Comment:
If there is only one key, maybe use `txn.Get()` is more efficient?

Reply:
Thanks for the suggestion, I will change it to `getKeyInTxn()` that does `txn.Get()` and checks if key was found or not.

@tangenta should I try to optimize the the `txn.BatchGet()` in `partitionedTableUpdateRecord()` below as well? I would assume that the normal/hot path would be `newFrom == newTo`, so it would only be a single key, otherwise two keys.

Reply:
I also optimized `partitionedTableUpdateRecord()` to do `txn.Get()`/`getKeyInTxn()` instead of `txn.BatchGet()`